### PR TITLE
[webgpu] Modularize ops

### PR DESCRIPTION
**Pull request by haoyunfeix**
FEATURE

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4594)

### CODE REVIEW

**Comment by haoyunfeix** [comment]
@qjia7 @lina128 @kainino0x PTAL

File: tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV5.ts [hunk 1]
```diff
@@ -14,38 +14,47 @@
...
 import {WebGPUBackend} from '../backend_webgpu';
 export type TypedArray = Float32Array|Int32Array|Uint8Array;
 
+export function nonMaxSuppressionV5(args: {
+  inputs: NonMaxSuppressionV5Inputs,
+  backend: WebGPUBackend,
+  attrs: NonMaxSuppressionV5Attrs
+}): [TensorInfo, TensorInfo] {
+  console.warn(
+      'tf.nonMaxSuppression() in webgl locks the UI thread. ' +
```
Comment (kainino0x) [inline tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV5.ts:28]: why this change from "webgpu" to "webgl"?

File: tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV3.ts [hunk 2]
```diff
@@ -15,26 +15,34 @@
...
-    console.warn(
-        'tf.nonMaxSuppression() in webgpu locks the UI thread. ' +
-        'Call tf.nonMaxSuppressionAsync() instead');
+export function nonMaxSuppressionV3(args: {
+  inputs: NonMaxSuppressionV3Inputs,
+  backend: WebGPUBackend,
+  attrs: NonMaxSuppressionV3Attrs
+}) {
+  console.warn(
+      'tf.nonMaxSuppression() in webgl locks the UI thread. ' +
```
Comment (kainino0x) [inline tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV3.ts:27]: same

**Review by kainino0x — APPROVED** [review]
As usual I'm unfamiliar with what modularized code in tfjs is supposed to look like, but this seems reasonable.

FYI to other reviewers, it's kind of necessary to review with "Hide whitespace changes" otherwise autoformatting changes crowd out actual changes.

**Review by haoyunfeix — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4594)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @kainino0x)

___
*[tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV3.ts, line 27 at r1](https://reviewable.io/reviews/tensorflow/tfjs/4594#-MRrOFp2sXT-h94Xftww-r1-27:-MSMddUNEh49Jw6TSPGo:bd5bo2h) ([raw file](https://github.com/tensorflow/tfjs/blob/03d6b466dc4ae9b300528ca92a8ebef1a83b729d/tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV3.ts#L27)):*
<details><summary><i>Previously, kainino0x (Kai Ninomiya) wrote…</i></summary><blockquote>

same
</blockquote></details>

Fixed typo, thanks!
___
*[tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV5.ts, line 28 at r1](https://reviewable.io/reviews/tensorflow/tfjs/4594#-MRrOFp2sXT-h94Xftwx-r1-28:-MSMdkpt45KrCvXkZfwR:bd5bo2h) ([raw file](https://github.com/tensorflow/tfjs/blob/03d6b466dc4ae9b300528ca92a8ebef1a83b729d/tfjs-backend-webgpu/src/kernels/NonMaxSuppressionV5.ts#L28)):*
<details><summary><i>Previously, kainino0x (Kai Ninomiya) wrote…</i></summary><blockquote>

why this change from "webgpu" to "webgl"?
</blockquote></details>

Fixed typo, thanks!

File: tfjs-backend-webgpu/src/kernels/Conv2D.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  ];
+
+  return backend.runWebGPUProgram(program, [x, filter], x.dtype, dimensions);
+}
+
+export const conv2DConfig: KernelConfig = {
+  kernelName: Conv2D,
+  backendName: 'webgpu',
+  kernelFunc: conv2d as {} as KernelFunc
+};
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/Conv2D.ts:80]: Add a new line at the end of the file.

File: tfjs-backend-webgpu/src/kernels/FusedConv2D.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  const {inputs, backend, attrs} = args;
+  const {x, filter, bias, preluActivationWeights} = inputs;
+  const {strides, pad, dataFormat, dilations, dimRoundingMode, activation} =
+      attrs;
+
+  const $dataFormat = backend_util.convertConv2DDataFormat(dataFormat);
+  const convInfo = backend_util.computeConv2DInfo(
+      x.shape as [number, number, number, number],
+      filter.shape as [number, number, number, number], strides, dilations, pad,
+      dimRoundingMode, false /* depthwise */, $dataFormat);
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/FusedConv2D.ts:40]: Below optimization is missing:
```
  if (convInfo.filterHeight === 1 && convInfo.filterWidth === 1 &&
      convInfo.dilationHeight === 1 && convInfo.dilationWidth === 1 &&
      convInfo.strideHeight === 1 && convInfo.strideWidth === 1 &&
      (convInfo.padInfo.type === 'SAME' || convInfo.padInfo.type === 'VALID')) {
    return conv2dByMatMul(	
          x, filter, convInfo, bias, preluActivationWeights, null,	
          activation, ...);
  }

**Review by haoyunfeix — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4594)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @kainino0x and @qjia7)

___
*[tfjs-backend-webgpu/package.json, line 24 at r3](https://reviewable.io/reviews/tensorflow/tfjs/4594#-MRrOFp2sXT-h94Xftwt-r3-24:-MSRrLGq0XjjljlouzOu:bb6un1j) ([raw file](https://github.com/tensorflow/tfjs/blob/b9a257b9fa87542ae1947c488626da9273fc9ff6/tfjs-backend-webgpu/package.json#L24)):*
<details><summary><i>Previously, qjia7 (Jiajia Qin) wrote…</i></summary><blockquote>

Please also exclude test `src/benchmark_models_test.ts` since it's for benchmark test.
</blockquote></details>

Done
___
*[tfjs-backend-webgpu/package.json, line 27 at r3](https://reviewable.io/reviews/tensorflow/tfjs/4594#-MRrOFp2sXT-h94Xftwt-r3-27:-MSRrOTbAoRQFpXB41mh:b61r3q6) ([raw file](https://github.com/tensorflow/tfjs/blob/b9a257b9fa87542ae1947c488626da9273fc9ff6/tfjs-backend-webgpu/package.json#L27)):*
<details><summary><i>Previously, qjia7 (Jiajia Qin) wrote…</i></summary><blockquote>

Should it also change to `yarn && yarn build-deps` for `benchmark`?
</blockquote></details>

Yes, fixed
___
*[tfjs-backend-webgpu/src/kernel_utils/shared.ts, line 1 at r3](https://reviewable.io/reviews/tensorflow/tfjs/4594#-MRrOFp2sXT-h94Xftwv-r3-1:-MSRrRoODVfHgZYovjI2:bb6un1j) ([raw file](https://github.com/tensorflow/tfjs/blob/b9a257b9fa87542ae1947c488626da9273fc9ff6/tfjs-backend-webgpu/src/kernel_utils/shared.ts#L1)):*
<details><summary><i>Previously, qjia7 (Jiajia Qin) wrote…</i></summary><blockquote>

Remove the extra blank line.
</blockquote></details>

Done
___
*[tfjs-backend-webgpu/src/kernels/Conv2D.ts, line 80 at r3](https://reviewable.io/reviews/tensorflow/tfjs/4594#-MRrOFp3oywkwKDdtv4E-r3-80:-MSRrABZ3KMUynRhBv4d:bb6un1j) ([raw file](https://github.com/tensorflow/tfjs/blob/b9a257b9fa87542ae1947c488626da9273fc9ff6/tfjs-backend-webgpu/src/kernels/Conv2D.ts#L80)):*
<details><summary><i>Previously, qjia7 (Jiajia Qin) wrote…</i></summary><blockquote>

Add a new line at the end of the file.
</blockquote></details>

Done
___
*[tfjs-backend-webgpu/src/kernels/FusedConv2D.ts, line 40 at r3](https://reviewable.io/reviews/tensorflow/tfjs/4594#-MRrOFp3oywkwKDdtv4I-r3-40:-MSRrCYj6rjMNy6jiLcM:b-oem2wb) ([raw file](https://github.com/tensorflow/tfjs/blob/b9a257b9fa87542ae1947c488626da9273fc9ff6/tfjs-backend-webgpu/src/kernels/FusedConv2D.ts#L40)):*
<details><summary><i>Previously, qjia7 (Jiajia Qin) wrote…</i></summary><blockquote>

Below optimization is missing:
```
  if (convInfo.filterHeight === 1 && convInfo.filterWidth === 1 &&
      convInfo.dilationHeight === 1 && convInfo.dilationWidth === 1 &&
      convInfo.strideHeight === 1 && convInfo.strideWidth === 1 &&
      (convInfo.padInfo.type === 'SAME' || convInfo.padInfo.type === 'VALID')) {
    return conv2dByMatMul(	
          x, filter, convInfo, bias, preluActivationWeights, null,	
          activation, ...);
  }
</blockquote></details>

Fixed, Thanks!


<!-- Sent from Reviewable.io -->

**Review by lina128 — COMMENTED** [review]
Great work! I only left comments about build process. Thanks.

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4594#-:-MSRriwS3UGsMlphwE6r:blhw7qp)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @haoyunfeix, @kainino0x, and @qjia7)

___
*[scripts/package_dependencies.json, line 13 at r3](https://reviewable.io/reviews/tensorflow/tfjs/4594#-MSRoG0m4jmAsorSSw_B:-MSRoG0m4jmAsorSSw_C:btsv77w) ([raw file](https://github.com/tensorflow/tfjs/blob/b9a257b9fa87542ae1947c488626da9273fc9ff6/scripts/package_dependencies.json#L13)):*
> ```JSON
>   "tfjs-backend-cpu": ["tfjs-core"],
>   "tfjs-backend-webgl": ["tfjs-core", "tfjs-backend-cpu"],
>   "tfjs-backend-webgpu": ["tfjs-core", "tfjs-backend-cpu", "tfjs-converter"],
> ```

package_dependencies.json is used to generate cloudbuild. The effect of adding this dependency means that every change in tfjs-core, tfjs-backend-cpu, and tfjs-converter will trigger webgpu test, this is for every PR and nightly. Given that webgpu is in active development, I would suggest separate its test from these regular and frequent builds. If we keep `tfjs-backend-webgpu: []` here, those changes won't trigger webgpu build. Only webgpu change or nightly build will trigger webgpu build. I suggest to do the necessary change in the cloudbuild.yml file in webgpu. So basically the steps should be `yarn` -> `yarn build-deps` -> `karma start`, take a look at webgl's cloudbuild.yml file, it should be something similar. Also, you probably don't need converter and webgl as your devDependency.

**Review by lina128 — CHANGES_REQUESTED** [review]
Please change the build process before merge.

**Review by haoyunfeix — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4594)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @kainino0x, @lina128, and @qjia7)

___
*[scripts/package_dependencies.json, line 13 at r3](https://reviewable.io/reviews/tensorflow/tfjs/4594#-MSRoG0m4jmAsorSSw_B:-MSWxUOJ-JTiezMG8FNa:b-8h4afu) ([raw file](https://github.com/tensorflow/tfjs/blob/b9a257b9fa87542ae1947c488626da9273fc9ff6/scripts/package_dependencies.json#L13)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

package_dependencies.json is used to generate cloudbuild. The effect of adding this dependency means that every change in tfjs-core, tfjs-backend-cpu, and tfjs-converter will trigger webgpu test, this is for every PR and nightly. Given that webgpu is in active development, I would suggest separate its test from these regular and frequent builds. If we keep `tfjs-backend-webgpu: []` here, those changes won't trigger webgpu build. Only webgpu change or nightly build will trigger webgpu build. I suggest to do the necessary change in the cloudbuild.yml file in webgpu. So basically the steps should be `yarn` -> `yarn build-deps` -> `karma start`, take a look at webgl's cloudbuild.yml file, it should be something similar. Also, you probably don't need converter and webgl as your devDependency. 
</blockquote></details>

Thanks for review and the explanation!

Yes, it's reasonable, I kept `tfjs-backend-webgpu: []` in this file and have to change `build-deps` to the name of `build-deps-for-webgpu` in cloudbuild.yml. Otherwise, generate_cloudbuild.js will skip these dependencies during CI building. I left a TODO in cloudbuild.yml to remind me go back to `build-deps` after webgpu backend is stable.

For `karma start`, I doubt there is any chrome canary in browserstack which could support webgpu to run `karma start`? So I kept only `yarn build` in test-ci. If there is a way to run chrome canary on browserstack please tell me.

And for convert and webgl dependency, I want to keep them there for now because we may use them frequently during our development for comparing with webgl backend and testing model benchmarks. Of course, I will remove them after webgpu backend is stable.

**Comment by qjia7** [comment]
@lina128 I am not familiar with the build process. Could you help double-check whether it addresses all your concerns?

**Review by lina128 — APPROVED** [review]
LGTM

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4594#-:-MSaLMtu29iTjtHUhJle:bb6ywmx)* status: :shipit: complete! 2 of 1 approvals obtained (waiting on @kainino0x, @lina128, and @qjia7)

**Comment by haoyunfeix** [comment]
@qjia7 Conflict resolved, PTAL

**Comment by qjia7** [comment]
Thanks, Yunfei. LGTM.

**Pull request merged**